### PR TITLE
[Inductor][CPP] Enhance cppcsevar data type deduce

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130317
* #122961
* #130892
* __->__ #130827

**Summary**
Previously, we used `data_type_propagation` at the start of `codegen` to deduce the data type of each node and save this information in `node.meta[OptimizationContext.key]`. Then, we used this node metadata to update the cppcsevar data type in `update_on_args`. However, this method is not always correct. For example, in the codegen of `indirect_indexing` (see [here](https://github.com/pytorch/pytorch/blob/096dc444cee3ebf8c1868ceb92ca4d6ede9f9cc1/torch/_inductor/codegen/common.py#L1844)), we insert nodes on the fly and reuse the node of `indirect_indexing` to set the `cppcsevar` data type. In this PR, we plan to enhance the `cppcsevar` data type deduction:

- We will deduce the `cppcsevar` data type in `update_on_args` by reusing the code in `data_type_propagation`.

- To align the data type of scalar and vector variables, we previously always cast the scalar to the vector's data type. This caused a data type misalignment between `codegen` and `data_type_propagation`. We should use the same data type promotion logic to align the data types of scalar and vector variables.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -535,22 +536,32 @@ def update_on_args(self, name, args, kwargs):
...
-            self.dtype = torch.bool
+        else:
+            # Induce output dtype by inputs' dtype
+            assert all(
+                arg.dtype is not None for arg in args if isinstance(arg, CppCSEVariable)
+            )
+            self.dtype = functools.reduce(
+                torch.promote_types,  # type: ignore[arg-type]
+                [arg.dtype for arg in args if isinstance(arg, CppCSEVariable)],
+            )
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:563]: Can we move this logic into `_deduce_output_dtype_by_name`?

File: torch/_inductor/codegen/cpp.py [hunk 2]
```diff
@@ -535,22 +536,32 @@ def update_on_args(self, name, args, kwargs):
...
+        ) is not None:
+            self.dtype = output_dtype
+        elif name == "masked":
+            assert (
+                hasattr(V.interpreter, "current_node")
+                and V.interpreter.current_node.target.startswith("masked_subblock")
+                and get_current_node_opt_ctx() is not None
+            )
             self.dtype = get_current_node_opt_ctx().dtype
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:554]: Perhaps just set the dtype per the current node context regardless of whether it is `masked` or not?

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -987,6 +991,52 @@ def sign(x):
...
+        functools.reduce(
+            torch.promote_types,  # type: ignore[arg-type]
+            [n.dtype for n in args if isinstance(n, CppCSEVariable)],
+        )
+        if all(n.dtype is not None for n in args if isinstance(n, CppCSEVariable))
+        else None  # not enough info to calculate the promote dtype
+    )
+
+
+def promote_vec_args(new_args):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:1005]: Shall we move them to cpp_utils.py?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:563]: Thanks for the comment. As synced offline, move it into a util function inside `cpp_utils.py`

File: torch/_inductor/codegen/cpp.py (hunk 2 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:554]: Perhaps we can also deduce the `masked` data type by inputs' cse variable of other. I have added a TODO here, let's check it later.

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:1005]: Sure, move it into cpp_utils.py.

File: torch/_inductor/codegen/common.py [hunk 4]
```diff
@@ -306,6 +306,47 @@ def boolean_ops():
 }
 
 
+def _deduce_output_dtype_by_name(
```
Comment (jgong5) [inline torch/_inductor/codegen/common.py:309]: It is used by other python files. Better to name it public.
```suggestion
def deduce_output_dtype_by_name(
```

File: torch/_inductor/codegen/cpp_utils.py [hunk 5]
```diff
@@ -76,6 +86,143 @@
...
+        functools.reduce(
+            torch.promote_types,  # type: ignore[arg-type]
+            [n.dtype for n in args if isinstance(n, CppCSEVariable)],
+        )
+        if all(n.dtype is not None for n in args if isinstance(n, CppCSEVariable))
+        else None  # not enough info to calculate the promote dtype
+    )
+
+
+def promote_vec_args(new_args):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_utils.py:100]: What's the diff from promoting normal `CppCSEVariable` args that could be either scalars or vectors?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/common.py (hunk 4 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/common.py:309]: Thanks for the comment and changed.

File: torch/_inductor/codegen/cpp_utils.py (hunk 5 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_utils.py]: yes, I think we can do the `promote_args` for `scalars` and `vectors` unified and changed.

**Review by jgong5 — APPROVED** [review]

**Comment by leslie-fang-intel** [comment]
Hi @lezcano, could you kindly help to take a look of this PR? thanks.

**Comment by leslie-fang-intel** [comment]
@jansel could you help to take a look of this PR?

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**